src/linter.cc

```diff
@@ -57,140 +59,162 @@ absl::Status PrintASTTree(absl::string_view sql) {
       break;
     }
   }
-  return return_status;
+  return LinterResult(return_status);
 }
 
-absl::Status CheckLineLength(absl::string_view sql, int line_limit,
+LinterResult CheckLineLength(absl::string_view sql, int line_limit,
                              const char delimeter) {
   int lineSize = 0;
   int line_number = 1;
+  int last_added = 0;
+  LinterResult result;
   for (int i = 0; i < static_cast<int>(sql.size()); ++i) {
     if (sql[i] == delimeter) {
       lineSize = 0;
       ++line_number;
     } else {
       ++lineSize;
     }
-    if (lineSize > line_limit) {
-      // TODO(orhanuysal): add proper error handling.
-      return absl::Status(
-          absl::StatusCode::kFailedPrecondition,
-          absl::StrCat("Lines should be <= ", std::to_string(line_limit),
-                       " characters long [", std::to_string(line_number),
-                       ",1]"));
+
+    if (lineSize > line_limit && line_number > last_added) {
+      last_added = line_number;
+      result.Add(ErrorCode::kLineLimit, sql, i,
+                 absl::StrCat("Lines should be <= ", std::to_string(line_limit),
+                              " characters long"));
     }
   }
-  return absl::OkStatus();
+  return result;
 }
 
-absl::Status CheckParserSucceeds(absl::string_view sql) {
+LinterResult CheckParserSucceeds(absl::string_view sql) {
   std::unique_ptr<ParserOutput> output;
 
   ParseResumeLocation location = ParseResumeLocation::FromStringView(sql);
 
   bool is_the_end = false;
+  int byte_position = 1;
+  LinterResult result;
   while (!is_the_end) {
-    ZETASQL_RETURN_IF_ERROR(ParseNextScriptStatement(&location, ParserOptions(),
-                                                     &output, &is_the_end));
+    byte_position = location.byte_position();
+    absl::Status status = ParseNextScriptStatement(&location, ParserOptions(),
+                                                   &output, &is_the_end);
+    if (!status.ok()) {
+      // TODO(orhanuysal): Implement a token parser to seperate statements
+      // currently, when parser fails, it is unable to determine
+      // the end of the statement.
+      result.Add(ErrorCode::kParseFailed, sql, byte_position, "Parser Failed");
+      break;
+    }
   }
-  return absl::OkStatus();
+  return result;
 }
 
-absl::Status CheckSemicolon(absl::string_view sql) {
+LinterResult CheckSemicolon(absl::string_view sql) {
+  LinterResult result;
   std::unique_ptr<ParserOutput> output;
 
   ParseResumeLocation location = ParseResumeLocation::FromStringView(sql);
   bool is_the_end = false;
 
   while (!is_the_end) {
-    ZETASQL_RETURN_IF_ERROR(ParseNextScriptStatement(&location, ParserOptions(),
-                                                     &output, &is_the_end));
-
+    absl::Status status = ParseNextScriptStatement(&location, ParserOptions(),
+                                                   &output, &is_the_end);
+    if (!status.ok()) return LinterResult(status);
```

Comment:
Actually, status returned by ParseNextScriptStatement contains information about error location, so you can convert it into a proper LinterResult. Create an issue to fix it later.

Reply:
Hmm storing it as a Status will ensure us enough information. They have their own Error type, so integration to class like LintError may not be necessary. We can just set a proper printing method, with lines and columns included.

Reply:
But then you sort LintErrors separately from Statuses, so the errors that came through a Status will not be sorted? Anyways, something to think about in a follow up improvements.